### DISCUSSION TITLE
Does Linq2Db cache complex query with inline parameters?

### DISCUSSION THREAD

**Discussion by timbze** [linq2db/linq2db#3931] [Q&A]
### Describe your issue

I have a complex query that seems to take around 130ms to generate sql every single time according to my logs (my computer is relatively fast).

Query is like:
```c#
_context.CustomerInfo
.Where(i => options.GetDisabled || !i.c.Disabled)
.OrderByDescending(i => !doSearch ? 0 :
    searchTerms.Sum(st =>
        st.WordSimilarityMaxAvg(i.c.Company, i.c.AdditionalSearchInfo)
    )
)
.ThenBy(i => i.c.Company)
.Skip(options.Skip)
.Take(options.Take)
.Select(i => i.c.MapToDomain(i.cm, i.n, i.p));
```

`WordSimilarity` is a Sql.Expression for PostgreSQL function `word_similarity`. The parameters are inlined in the output SQL. I did also try changing it a bit so that parameters were not inline, but it did not seem to make a difference in speed.

I do have a few joins that are removed for brevity.

Do you know if this query would be cached in Linq2Db? If not, how can I change it to be generated faster? How can I find out if the query is in cache?

### Environment details

`Linq To DB` version: *5.0.0 preview 1*

Database (with version): *PostgreSQL 14.5*

Operating system: *Ubuntu 22.04*

.NET Version: *7.0*

**Comment by MaceWindu** [comment]
Should be cached. You can check that `Query<YourResultType>.CacheMissCount` is not increased after each query.
If you provide mappings we can look at this query in details

**Reply by timbze** [reply]
I have the weirdest issue, no idea how it's even possible. Thanks for the info on how I can find if Query is cached.

I use [a library](https://github.com/TecMeet/tecmeet-openiddict-linq2db) I created/adapted, and as soon as I add [`UseOpenIddict`](https://github.com/TecMeet/tecmeet-openiddict-linq2db/blob/65008ea79da43368d0f7abe24b74bc4c93364c04/src/TecMeet.OpenIddict.LinqToDB/OpenIddictLinqToDBExtensions.cs#L113) I get cache misses for each new scope. While 1 API endpoint is running there will only be one miss even with multiple db calls, but as soon as there's a new API call there will again be another miss.

I can do everything manually in my project that `UseOpenIddict` does and everything is perfect. But once I call that method from my library I get cache misses.

This does NOT work:
```c#
services.AddLinqToDBContext<DbContext>((serviceProvider, opt) => opt
    .UseConnectionFactory(PostgreSQLTools.GetDataProvider(PostgreSQLVersion.v95), ds.CreateConnection)
    .UseOpenIddict<Guid>(ms)
    .UseMappingSchema(ms)
    .UseDefaultLogging(serviceProvider));
```

This works:
```c#
services.AddLinqToDBContext<DbContext>((serviceProvider, opt) => opt
    .UseConnectionFactory(PostgreSQLTools.GetDataProvider(PostgreSQLVersion.v95), ds.CreateConnection)
//    .UseOpenIddict<Guid>(ms)
    .UseMappingSchema(ms)
    .UseDefaultLogging(serviceProvider));
```

This also works (do myself what UseOpenIddict does):
```c#
ms.GetFluentMappingBuilder()
    .Entity<OpenIddictLinqToDBApplication<Guid>>()
    .HasTableName("auth_openiddict_applications")
    .Property(i => i.Id).IsPrimaryKey().IsIdentity()
    .Entity<OpenIddictLinqToDBAuthorization<Guid>>()
    .HasTableName("auth_openiddict_authorizations")
    .Property(i => i.Id).IsPrimaryKey().IsIdentity()
    .Entity<OpenIddictLinqToDBScope<Guid>>()
    .HasTableName("auth_openiddict_scopes")
    .Property(i => i.Id).IsPrimaryKey().IsIdentity()
    .Entity<OpenIddictLinqToDBToken<Guid>>()
    .HasTableName("auth_openiddict_tokens")
    .Property(i => i.Id).IsPrimaryKey().IsIdentity();

services.AddLinqToDBContext<DbContext>((serviceProvider, opt) => opt
    .UseConnectionFactory(PostgreSQLTools.GetDataProvider(PostgreSQLVersion.v95), ds.CreateConnection)
//    .UseOpenIddict<Guid>(ms)
    .UseMappingSchema(ms)
    .UseDefaultLogging(serviceProvider));
```

Any idea how this can be?

**Reply by timbze** [reply]
I have now changed the library to only use the MappingSchema since the DataOptions is no longer needed.

Now I do this and it works:
```c#
ms.UseOpenIddict<Guid>();

services.AddLinqToDBContext<DbContext>((serviceProvider, opt) => opt
    .UseMappingSchema(ms)
    .UseConnectionFactory(PostgreSQLTools.GetDataProvider(PostgreSQLVersion.v95), ds.CreateConnection)
    .UseDefaultLogging(serviceProvider));
```

**Reply by timbze** [reply]
FYI, the lib with the issue was 2.0.0-rc1. The updated lib is 2.0.0-rc2

**Comment by MaceWindu** [comment]
I've checked `UseOpenIddict` extension code - your fix is correct. It adds additional mappings to mapping schema using fluent mappings and it's not something you want to do on each context creation. Mapping schema should be configured only once.
I would even recommend to use single `DataOptions` instance to avoid creation of it for each connection instance, but it's not so imprtant as mapping schema.

**Reply by timbze** [reply]
Thank you for the advice!

How would you use single instance of `DataOptions` since I'm also using the `ServiceProvider` for the `UseDefaultLogging`?

I probably could use the `IServiceCollection.BuildServiceProvider()` but I don't know if that is the right way (or good way) to go.

**Reply by MaceWindu** [reply]
As you use `UseDefaultLogging` which require `serviceProvider` best you can do is to preconfigure all other options only once:

```cs
var staticOptions = new DataOptions()
    .UseMappingSchema(ms)
    .UseConnectionFactory(PostgreSQLTools.GetDataProvider(PostgreSQLVersion.v95), ds.CreateConnection);

services.AddLinqToDBContext<DbContext>((serviceProvider, opt) => staticOptions.UseDefaultLogging(serviceProvider));
```

**Reply by timbze** [reply]
This seems to work fine. Do you know of a reason why it would be bad to do?

```c#
var dataOptions = new DataOptions()
    .UseConnectionFactory(PostgreSQLTools.GetDataProvider(PostgreSQLVersion.v95), ds.CreateConnection)
    .UseMappingSchema(ms)
    .UseDefaultLogging(services.BuildServiceProvider());

services.AddLinqToDBContext<IDbContext, DbContext>((_,_) => dataOptions);
```

**Reply by MaceWindu** [reply]
Don't see any problem here

**Reply by timbze** [reply]
Thank you very much for the help!
